### PR TITLE
Fix retry_obj retryMutex vs retryEntry.Mutex deadlocks

### PR DESCRIPTION
**Pull request by npinaeva**
Implmenet syncCache for retryObjects with lockable keys.

Isolate cache implementation from the retry logic:
  - retryObjEntry doesn't need to have mutex anymore
  - retryObjs doesn't work with cache mutexes, only calls methods

SyncCache is a map with lockable keys. It allows to lock the key
regardless of whether the entry for given key exists.
When key is locked other threads can't read/write the entry with
the key.

Split ensureRetryEntryLocked into loadOrStore that will
make sure newEntry is present and locked, and load that was
previously called as ensureRetryEntryLocked with nil newRetryEntry.
Change iterateRetryResources to only lock every entry once and not use
cache mutex. Create a snapshot of keys that will be retried instead
of holding map lock.

Lock retyrObject key until work for this key is completed.
That also removes the need for .ignore field - retry loop
will wait until key is unlocked.

Signed-off-by: Nadia Pinaeva <npinaeva@redhat.com>

### CODE REVIEW

**Review by trozet — CHANGES_REQUESTED** [review]
Nadia pointed out there are more places where deadlock can happen:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L1255-L1261

https://github.com/ovn-org/ovn-kubernetes/blob/3d098924fca5020805c08546f78cfd7d29304129/go-controller/pkg/ovn/obj_retry.go#L239-L242

I would prefer we stay consistent with how we handle maps and locking of entries by following the same strategy we use for egress gw and nsInfo:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L59

https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/namespace.go#L408

That way we can be consistent in the code and leverage something we already know works (plus we spent a ton of time getting those to work correctly)

File: go-controller/pkg/ovn/obj_retry.go [hunk 1]
```diff
@@ -170,94 +171,134 @@ func (r *retryObjs) initRetryObjWithDelete(obj interface{}, key string, config i
...
+		// entry exists, try to Lock it
 		entry.Lock()
+		// check if entry was deleted while we waited for lock
+		// and add newRetryEntry if that's the case
+		updatedEntry, loaded := r.loadOrStoreRetryEntry(key, newRetryEntry)
+		if loaded {
+			// entry we loaded and locked is still in the map, return it
+			// unlock our tmp newRetryEntry, since we didn't use it
+			newRetryEntry.Unlock()
+			return entry
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go:293]: That's a good observation - and all 3 files (egressgw, namespace and old ensureRetryEntryLocked) implement it differently. I like the egressgw option with .deleted field, but we need to create a single interface - because what we have now is not reusing the code, but copying it - which as we can see results in having 3 different implementations
egressgw uses .deleted https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L88
namespace compares objects https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/namespace.go#L449
and old ensureRetryEntryLocked ignores this (as current PR) https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L243

**Review by trozet — COMMENTED** [review]
I think we need a better way of locking so that add/delete operations from the cache are guaranteed to work. I think that means we need to lock the cache before locking entries. Perhaps this will work:
Add:
lock cache
lock entry
if entry.deleted, unlock entry and return nil
unlock cache, and return entry

Get
lock cache
lock entry
if entry.deleted, unlock entry and return nil
unlock cache and return entry

Delete
mark as deleted
unlock entry
lock cache
lock entry
delete entry from cache
unlock entry
unlock cache

This way we should avoid dead lock where we have entry.Lock and we are trying to get cache.Lock. Maybe @dcbw has other ideas

**Comment by npinaeva** [comment]
@trozet your comments are vary valid (thanks for reviewing) - and I think based on them I figured out the right semantics for this cache.
We don't want to lock the entry, we want to lock the key. Meaning when a thread access entry with some key nobody else should work with the same key until the first thread releases the key lock. This also works when there is no entry in the cache, key lock still will be acquired and nobody else will be able to add the entry until key lock is released. Same works for deletion - while key lock is acquired nobody can delete existing object from the cache. 

It looks like this allows us to guarantee no errors for cache operations (please review SyncCache carefully), but I think we need to rework retry_obj with this new approach, because now the code doesn't care what happens with the entry e.g. between create and update operations https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L1429-L1436, https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L1323-L1324, (there are other examples) but we can fix that with the key lock by releasing the lock when all the changes for this entry are done.

File: go-controller/pkg/ovn/obj_retry.go [hunk 2]
```diff
@@ -1191,7 +1306,21 @@ func (oc *Controller) resourceRetry(r *retryObjs, lre *localRetryEntry, now time
...
+				klog.Infof("%s %s not found in the informers cache,"+
+					" not going to retry object create", r.oType, objKey)
+				kObj = nil
+			} else {
+				klog.Errorf("Failed to look up %s %s in the informers cache,"+
+					" will retry later: %v", r.oType, objKey, err)
+				return
+			}
+		}
+		entry.newObj = kObj
```
Comment (jcaamano) [inline go-controller/pkg/ovn/obj_retry.go:1323]: This change seems unrelated, can you split it to a different commit and explain?

File: go-controller/pkg/ovn/egressgw.go [hunk 3]
```diff
@@ -392,6 +392,9 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod *kapi.Pod, namespace str
 	if err := oc.deleteGWRoutesForNamespace(namespace, foundGws.gws); err != nil {
 		// add the entry back to nsInfo for retrying later
 		nsInfo, nsUnlock := oc.getNamespaceLocked(namespace, false)
+		if nsInfo == nil {
+			return fmt.Errorf("failed to get nsInfo %s to add back all the gw routes: %w", namespace, err)
+		}
```
Comment (jcaamano) [inline go-controller/pkg/ovn/egressgw.go:397]: A few lines above if nsInfo == nil we return nil, but here we return an error. If this indeed should be the case, I am missing some comments that explain.

File: go-controller/pkg/ovn/namespace.go [hunk 4]
```diff
@@ -359,7 +359,7 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 			klog.Errorf("Failed to delete network policy: %s, error: %v", key, err)
 			oc.retryNetworkPolicies.unSkipRetryObj(key)
 		} else {
-			oc.retryNetworkPolicies.deleteRetryObj(key, true)
+			oc.retryNetworkPolicies.retryEntries.Delete(key, false)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/namespace.go:362]: This feels like it should read `oc.retryNetworkPolicies.Delete(key)`

File: go-controller/pkg/ovn/obj_retry.go [hunk 5]
```diff
@@ -53,10 +52,191 @@ type retryObjEntry struct {
...
+		sync.Mutex{},
+		sync.Mutex{},
+		0,
+	}
+	return &c
+}
+
+// SyncCache is a map with lockable keys. It allows to lock the key regardless of whether the entry for
+// given key exists. When key is locked other threads can't read/write the entry with the key.
+type SyncCache struct {
```
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/obj_retry.go:84]: If you have any concerns using ObjEntryKeyLock as defined in my other comment above, you can use something else
`type SomeOtherSpecificKeyType string`

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/egressgw.go:397]: I think the problem is github showing too less lines - this `err != nil` branch ends up returning an error after modifying `nsInfo` like 4 lines lower. So this change just prevents nil dereference and returns without adding entry back.
let me know if you think I can add any comments to the code to explain this better

File: go-controller/pkg/ovn/obj_retry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go:84]: This is genius! This interface looks much better! I am leaving only the methods that are used for now (e.g. LoadAndDelete is not implemented yet), but required methods can be added later.

about moving to its own package - I agree and I have a new branch where SyncCache is generic (requires to update our go to 1.18) and lives in utils, but here it uses `retryObjEntry` so I decide to leave it here for now and do the mentioned changes later.

File: go-controller/pkg/ovn/obj_retry.go (hunk 2 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go:1323]: It is related - this part was modes from `iterateRetryResources` https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L1263-L1276 to reduce the need for additional entry lock - now it is a part of `resourceRetry`

File: go-controller/pkg/ovn/obj_retry.go [hunk 6]
```diff
@@ -1134,31 +1181,19 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
-}
-
-func (oc *Controller) resourceRetry(r *retryObjs, lre *localRetryEntry, now time.Time) {
-	objKey := lre.key
-	entry := r.ensureRetryEntryLocked(objKey, nil)
-	if entry == nil {
+func (oc *Controller) resourceRetry(r *retryObjs, objKey string, now time.Time) {
+	lockedKey := r.lockKey(objKey)
+	defer r.unlockKey(lockedKey)
+	entry, loaded := r.getRetryObj(lockedKey)
```
Comment (trozet) [inline go-controller/pkg/ovn/obj_retry.go:1187]: There is a relationship between keyLocks and entries, where no one should modify an entry without a lock for the same key in keyLocks map. However, this is not enforced. I think we should create a common ExecuteWithLock(key, func) that will lock the key and execute whatever on the same keyed entry. In the future once this is moved to its own library, a user will not have access to the entries map directly. I'm worried about bugs where a developer accidentally modifies an entry without a key lock.

**Review by trozet — COMMENTED** [review]
Seems like this will work, but I'm hesitant about 2 maps with more mutexes. I ran sasha deadlock on it and it was clean.

File: go-controller/pkg/ovn/obj_retry.go [hunk 7]
```diff
@@ -45,18 +44,174 @@ type retryObjEntry struct {
...
+		// and when the key is Locked, c.keyLocks[key] will always have its keyLock.
+		// similar to calling Unlock() on unlocked mutex
+		klog.Errorf("Unlocking non-existing key %s", key)
+		return
+	}
+	kLock.delRef()
+	// keyLock can be deleted when the last request is being shut down (that is refCount=0)
+	// next load request for this key will create a new keyLock
+	if kLock.refCounter == 0 {
+		c.entriesMutex.Lock()
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go:121]: maps are not safe for concurrent use https://go.dev/doc/faq#atomic_maps
and this map will be accessed concurrently for different keys

File: go-controller/pkg/ovn/obj_retry.go (hunk 6 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go:1187]: Actually it is enforced (that's why we added `type LockedKey string`): entries are not accessed directly by anything except SyncCache, and all SyncCache methods that access entries look like
```
func (c *SyncCache) Load(key LockedKey) (value *retryObjEntry, loaded bool)
func (c *SyncCache) LoadOrStore(key LockedKey, newEntry *retryObjEntry) *retryObjEntry
func (c *SyncCache) Delete(key LockedKey)
```
they have `LockedKey` as an argument and you can only get it by calling 
```
func (c *SyncCache) LockKey(key string) LockedKey
```
So the usual workflow will be like
```
lockedKey := oc.retrySmth.lockKey(key)
defer oc.retrySmth.unlockKey(lockedKey)
// do something like
oc.retrySmth.initRetryObjWithDelete(np.policy, lockedKey, np, false)
isLastPolicyInNamespace := len(nsInfo.networkPolicies) == 1
if err := oc.destroyNetworkPolicy(np, isLastPolicyInNamespace); err != nil {
	klog.Errorf("Failed to delete network policy: %s, error: %v", key, err)
} else {
	oc.retryNetworkPolicies.deleteRetryObj(lockedKey)
	delete(nsInfo.networkPolicies, np.name)
}

```

File: go-controller/pkg/ovn/obj_retry.go [hunk 8]
```diff
@@ -93,73 +247,53 @@ func NewRetryObjs(
...
+	return r.retryEntries.LockKey(key)
 }
 
-// initRetryObjWithAdd creates a retry entry for an object that is being added,
-// so that, if it fails, the add can be potentially retried later.
-// initially it is marked as skipped for the retry loop (ignore = true).
-func (r *retryObjs) initRetryObjWithAdd(obj interface{}, key string) {
-	r.initRetryObjWithAddBackoff(obj, key, initialBackoff)
+func (r *retryObjs) unlockKey(key LockedKey) {
+	r.retryEntries.UnlockKey(key)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/obj_retry.go:255]: I think we should avoid leaking the concept of a locked key outside of retryObjs, at least on this form, but this is a bit of a mess, there is no clear interface of what is meant to be private to retryObjs (or testing only) or public to the outside world, and this is because we are not moving this to it's own package. 

Also unclear on what is the interaction between Controller and retry_objs, they seem to be tightly coupled, but at least it is hopefuly confined within this file.

So my suggestion is, for now:

- Capitalize methods meant to be used outside of retryObjs and do no leak the concept of locked key there:
  * AddRetryObj* 
  * DeleteRetryObj*
  * ...
-  keep lower case for other private methods meant to be used within retryObjs (or controller methods within this file) that need a locked key
   * initRetryObj*
   * ...
- move testing methods that you have in pod_test.go but you use everywhere to obj_retry_test.go 

I would remove these lock/unlock methods once it's clear they are only need privately, within obj_retry we can just use r.retryEntries.LockKey, at least until this is moved to it's own package.

File: go-controller/pkg/ovn/namespace.go [hunk 9]
```diff
@@ -350,18 +350,19 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
...
 		// along with any mappings of nsInfo -> network policies
-		oc.retryNetworkPolicies.initRetryObjWithDelete(np.policy, key, np, false)
+		oc.retryNetworkPolicies.initRetryObjWithDelete(np.policy, lockedKey, np, false)
 		isLastPolicyInNamespace := len(nsInfo.networkPolicies) == 1
 		if err := oc.destroyNetworkPolicy(np, isLastPolicyInNamespace); err != nil {
 			klog.Errorf("Failed to delete network policy: %s, error: %v", key, err)
-			oc.retryNetworkPolicies.unSkipRetryObj(key)
 		} else {
-			oc.retryNetworkPolicies.deleteRetryObj(key, true)
+			oc.retryNetworkPolicies.deleteRetryObj(lockedKey)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/namespace.go:362]: Have a `AddRetryObjWithDelete` and `DeleteRetryObject` in retryObjs that handles key locking internally

File: go-controller/pkg/ovn/policy.go [hunk 10]
```diff
@@ -1320,8 +1320,9 @@ func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) error {
 		if err := oc.deleteNetworkPolicy(policy, np); err != nil {
 			// rollback failed, add to retry to cleanup
 			key := getPolicyNamespacedName(policy)
-			oc.retryNetworkPolicies.initRetryObjWithDelete(policy, key, np, false)
-			oc.retryNetworkPolicies.unSkipRetryObj(key)
+			lockedKey := oc.retryNetworkPolicies.lockKey(key)
+			oc.retryNetworkPolicies.initRetryObjWithDelete(policy, lockedKey, np, false)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:1324]: Have a AddRetryObjectWithDelete in retryObjs that handles key locking internally

File: go-controller/pkg/ovn/pods_test.go [hunk 11]
```diff
@@ -216,6 +216,64 @@ func getExpectedDataPodsAndSwitches(pods []testPod, nodes []string) []libovsdbte
...
+		expectedValue = gomega.BeFalse()
+	}
+	gomega.Eventually(func() bool {
+		return checkRetryObj(key, r)
+	}).Should(expectedValue)
+}
+
+func retryObjsLen(r *retryObjs) int {
+	return len(r.retryEntries.GetKeys())
+}
```
Comment (jcaamano) [inline go-controller/pkg/ovn/pods_test.go:275]: Does it make sense to move these methods to obj_retry_test.go?

File: go-controller/pkg/ovn/obj_retry.go (hunk 6 above)
Comment (jcaamano) [reply go-controller/pkg/ovn/obj_retry.go:1187]: Right, LockedKey is something the SyncCache exports to SyncCache users, but that does not meant retryObjs should export the same thing to retryObjs users as happened in the last push. But this is all a bit confusing since we are throwing everything in with no clear drawn line between interfaces.

I would not leak the concept of locked key outside of objs_retry.go, SyncCache and SyncCache direct users until we put some order here.

File: go-controller/pkg/ovn/obj_retry.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/obj_retry.go:121]: well I know that...but you aren't accessing the entries map, you are accessing the keyLocks map which you already have the mutex for. So why do you need to lock the entries map to delete something from the keylocks map?

**Comment by npinaeva** [comment]
Let me try to explain the concept of keys locking a bit more - in contrast from locking the object, key locking should be explicit (that requires using `LockedKey` type). Using it explicitly allows us to 
1. Explicitly control when the key is locked and unlocked on a handler level - handlers need to know that nobody touches the key while they are doing some changes. @jcaamano you suggest wrapping functions with lock, but I would argue that having a unified interface for retryObjs, where retryObjs itself never locks the key with its methods is more clean - that is the whole point of keys locking - handler does that. And e.g. here https://github.com/ovn-org/ovn-kubernetes/blob/dc3a8de59b9f5de5de293de37027d3cd42acf5e0/go-controller/pkg/ovn/namespace.go#L353-L365 if I wrap add and delete retryObj with locks I will have to lock the same objects twice and there are to guarantees that nothing happens with it in between those calls. That is error-prone
2. For methods/function - require the key to be locked via its signature with `LockedKey`
3. For consumers - not worry about wrongly lock/unlock key - if the method works with the entry it will require `LockedKey` as an argument, if `LockedKey` was received it should be unlocked
4. Interfaces become simpler - main handler locks and unlocks key when needed - all other methods (like `initRetryObj` and all methods that modify object) don't need to care about key locking - they work with  `retryEntries` as with simple sync.Map

From the isolation perspective - I moved syncMap to a separate package, you can check that syncMap fields are not accessed from the outside, so the interface it provides is `LockedKey` and its methods.
retryObjs also only uses its internal retryEntries (same as all its fields) by itself - that's why I have methods like 
```
func (r *retryObjs) LockKey(key string) syncmap.LockedKey {
	return r.retryEntries.LockKey(key)
}
```
About renaming retryObjs methods - it's a bit messy, because one `initRetryObj` may be only used locally and another one may be used from other files, so I renamed some method to Upper-case if they are used externally or belong to the same "logical" group of externally called methods.

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/egressgw.go:397]: If `oc.deleteGWRoutesForNamespace(namespace, foundGws.gws)` failed, it means cleanup wasn't completed and needs to be retried - so this `if` branch should return error anyway. We can't just hope that whoever deleted nsInfo also did the proper cleanup right?

File: go-controller/pkg/ovn/obj_retry.go (hunk 6 above)
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go]: I think it is true - if every key handling includes only 1 operation with this key, all key locking interface can be hidden, but that is not true for retry_obj algorithms - they get, create, update, delete one key just during update event hendling

**Comment by npinaeva** [comment]
I would also suggest in the future to unify `WatchResource` and `resourceRetry` (because they should be doing the same thing) so that every new event just needs to create retry entry with no backoff and leave retry loop to take care of it - then most of handler logic will look like 
lock key
create retry obj
unlock key
and here we can change RetryObjs interfaces to wrap lock calls and only expose `InitRetryObj` without any LockedKey

**Comment by jcaamano** [comment]
>  @jcaamano you suggest wrapping functions with lock, but I would argue that having a unified interface for retryObjs, where retryObjs itself never locks the key with its methods is more clean - that is the whole point of keys locking - handler does that. 

> if I wrap add and delete retryObj with locks I will have to lock the same objects twice and there are to guarantees that nothing happens with it in between those calls. That is error-prone

It's unusual to have an exported interface with a type that comes from another privately used type (LockedKey as argument in exported methods of retryObjs). What if we decide to change this SyncMap with something else in the future that solves the problem better? Then we would need to change retryObjs interface and the handlers implementation as well. This is not a end-of-the-world problem but certainly a improvable design and certainly stood up to me when I saw it.

Moreover, alternatives for the retryObjs interface with different semantics but the same dynamics could also be better than LockKey/UnlockKey. Example `DoWithLockedEntry(key, func(entry))` that @trozet suggested could work better. But I don't think we need anything at all now, all this locking business can stay private inside obj_retry.go and **entries** or **anything sensible to this locking semantic** should be kept inside until we take what I think is the most important step towards safe retry entry operation which is to actually move retryObjs to it's own package.

I think that bumping 1.18 and dealing with the novelty of generics might delay this PR more than necessary as well. Now I would have to argue why the key are strings instead of generic :P

**Comment by npinaeva** [comment]
alright - here is what I came up with:
1. Remove LockedKey type - syncMap doesn't protect users from calling methods without key lock, but I think its interface is clear enough to use it properly (I also renamed func args to `lockedKey string` to make it even more obvious)
2. Responsibility for proper key locking is on the syncMap users - i.e. RetryObjs. I added `DoWithLock` method for it that locks the key and all the other operation look like a simple map. I left "exported" only methods that lock the key (DoWithLock, AddRetryObjWithAddNoBackoff) all other "local" methods should be called inside DoWithLock. I also renamed func args that should be called with lock to `lockedKey string`
3. About using generics and go 1.18 - for me it doesn't look like a big change (and I think we can agree to use string for syncCache since there are no other use cases for now - and we don't want to overengineer it). If bumping go version looks dangerous I can return syncMap to retry_obj file, but I think having it in a separate package makes things more clear.

File: go-controller/pkg/ovn/obj_retry.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/obj_retry.go]: lol I read that wrong - I don't need it here of course, thanks for catching :)

**Comment by npinaeva** [comment]
/retest
```
Summarizing 1 Failure:

[Fail] OVN NetworkPolicy Operations during execution ACL logging for network policies [It] policies created after namespace logging level updates inherit updated logging level 
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/policy_test.go:2872
``` 
doesn't looks like this PR problem

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (trozet) [reply go-controller/pkg/ovn/egressgw.go:397]: I think returning error is fine here. If the namespace is gone, then when this is retried deleteGWRoutesForNamespace will not return error, and we wouldn't enter this block again.

**Comment by npinaeva** [comment]
catching new unit test failure while fixing another one - I'm lovin' it
Here should be the fix - I decided to add it to another PR I have around the same lines https://github.com/ovn-org/ovn-kubernetes/pull/3093/commits/0c5404424b30d5c0c4cdea89f1fa194d1563eea4

**Review by trozet — APPROVED** [review]
would like to see some unit tests on the syncmap but we can do that later. @jcaamano PTAL

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (jcaamano) [reply go-controller/pkg/ovn/egressgw.go:397]: Ah, ok, thanks for explaining, I wasn't seeing the full picture.

File: go-controller/pkg/ovn/obj_retry.go [hunk 12]
```diff
@@ -1472,51 +1333,49 @@ func (oc *Controller) WatchResource(objectsToRetry *retryObjs) (*factory.Handler
...
-						klog.Errorf("Failed to delete old object %s of type %s,"+
-							" during add event: %v", key, objectsToRetry.oType, err)
+					retryObj := objectsToRetry.initRetryObjWithAdd(obj, key)
+					// If there is a delete entry with the same key, we got an add event for an object
+					// with the same name as a previous object that failed deletion.
+					// Destroy the old object before we add the new one.
+					//
+					// Note it is okay to access retryEntry without lock here, as the entry is either
+					// accessed by iterateRetryResources or in add/update/delete handler of this resource;
+					// the former is prevented as its ignore field is set to true, and the latter is serialized.
```
Comment (jcaamano) [inline go-controller/pkg/ovn/obj_retry.go:1351]: Is this comment still relevant?

**Comment by jcaamano** [comment]
> alright - here is what I came up with:
> 
> 1. Remove LockedKey type - syncMap doesn't protect users from calling methods without key lock, but I think its interface is clear enough to use it properly (I also renamed func args to `lockedKey string` to make it even more obvious)

I did not have a problem with this (it was rather with having it in RetryObjs methods) but it is fine as it is now as well.

> 2. Responsibility for proper key locking is on the syncMap users - i.e. RetryObjs. I added `DoWithLock` method for it that locks the key and all the other operation look like a simple map. I left "exported" only methods that lock the key (DoWithLock, AddRetryObjWithAddNoBackoff) all other "local" methods should be called inside DoWithLock. I also renamed func args that should be called with lock to `lockedKey string`

Looks better now. This is almost ready to be moved to it's own package. `DoWithLockedEntry(key string, func(*entry))` and then we can gate what you need to do with a locked key requiring to have an entry, instead of being able to do it with a key. Anyway, not needed now so food for later thought. Happy with how it looks like now.

> 3. About using generics and go 1.18 - for me it doesn't look like a big change (and I think we can agree to use string for syncCache since there are no other use cases for now - and we don't want to overengineer it). If bumping go version looks dangerous I can return syncMap to retry_obj file, but I think having it in a separate package makes things more clear.

Fair enough. The alternative was to use interface{} and just make the proper cast in getRetryObj or something like that.

So this looks fine with me. I got a bit lost reviewing the handlers and resourceRetry, what an unfortunate diff display but I guess it did not change as much as it looks like and hopefully @trozet has the handle of that as well.

File: go-controller/pkg/ovn/obj_retry.go (hunk 12 above)
Comment (trozet) [reply go-controller/pkg/ovn/obj_retry.go:1351]: dont think so but we can clean it up later
